### PR TITLE
Move syntax checks about lambda-parameter type mismatches out of the parser.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler @jcouv @cston

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz Improvements in parser that benefit IDE and compiler.

**Comment by MattGertz** [comment]
Approved pending tests.

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by agocke** [comment]
\+ @jaredpar Does this meet our bar? Is the bar different between the IDE and the compiler? If we change both do we have to get double-ask-mode approval?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -10828,59 +10828,60 @@ private AnonymousMethodExpressionSyntax ParseAnonymousMethodExpression()
...
+            if (this.CurrentToken.ContextualKind == SyntaxKind.AsyncKeyword &&
+                PeekToken(1).Kind != SyntaxKind.EqualsGreaterThanToken)
             {
                 asyncToken = this.EatContextualToken(SyntaxKind.AsyncKeyword);
                 asyncToken = CheckFeatureAvailability(asyncToken, MessageID.IDS_FeatureAsync);
                 IsInAsync = true;
             }
 
-            ExpressionSyntax result;
+            var result = ParseLambdaExpression(asyncToken);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I think our convention is to have spelled out type here. Alternatively, you can probably swap this line and the next, and directly return without the `result` local.

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Nevermind on spelled out type. I see the rest of this method was already using `var` ;-)

---
In reply to: [112343049](https://github.com/dotnet/roslyn/pull/16419#discussion_r112343049) [](ancestors = 112343049)

File: src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs [hunk 2]
```diff
@@ -180,6 +181,32 @@ internal partial class Binder
             return Tuple.Create(refKinds, types, names, isAsync);
         }
 
+        private void CheckParanthesizedLambdaParameters(
+            SeparatedSyntaxList<ParameterSyntax> parameterSyntaxList, DiagnosticBag diagnostics)
+        {
+            if (parameterSyntaxList.Count > 0)
+            {
+                var hasTypes = parameterSyntaxList[0].Type != null;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Lambda.cs]: I understand this preserves the previous behavior, but it raises the question: should we check all the parameters for presence of type?
(if yes, let's file an issue)

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs [hunk 3]
```diff
@@ -4066,7 +4066,7 @@ public class Test
             // TODO...
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/18800")]
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: I see that the issue says this was fixed by Andy. Should we unskip now?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: Yes.  I have a bug on me to unskip :)

**Comment by CyrusNajmabadi** [comment]
Rebased to 15.6

**Comment by CyrusNajmabadi** [comment]
CSharpNavigateTo.NavigateTo is tracked by https://github.com/dotnet/roslyn/issues/19530

**Pull request merged**